### PR TITLE
Eliminate closure allocations in System.Net.Http

### PR DESCRIPTION
**Pull request by Maxwe11**

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/HttpClient.cs [hunk 1]
```diff
@@ -156,41 +156,124 @@ public Task<Stream> GetStreamAsync(string requestUri)
...
                 content => content.ReadAsStreamAsync());
         }
 
-        private Task<T> GetContentAsync<T>(Uri requestUri, HttpCompletionOption completionOption, T defaultValue,
-            Func<HttpContent, Task<T>> readAs)
+        private Task<string> GetStringContentAsync(Uri requestUri, HttpCompletionOption completionOption, 
+            Func<HttpContent, Task<string>> readAs)
         {
-            TaskCompletionSource<T> tcs = new TaskCompletionSource<T>();
+            var tcs = new TaskCompletionSource<string>(readAs);
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/HttpClient.cs]: This ```readAs``` delegate is now going to be exposed from the returned ```Task```'s AsyncState property.  That's probably undesirable, but it's up to @davidsh et al whether they're comfortable with that.

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/HttpClient.cs]: (I'll refrain from reviewing the rest of the PR until that's commented on, as is it appears most of the change relies on passing state through in that fashion.  If the team is ok with it, I am, too.)

**Comment by davidsh** [comment]
@Maxwe11 @stephentoub We'll respond back in a day or so.  We might want to pilot this change in our full end-to-end tests which haven't yet been put on GitHub.  Also,need to finish #2149 before that.

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by Maxwe11** [comment]
@dotnet-bot test this please

**Comment by davidsh** [comment]
@stephentoub Regarding your comment: 

>(I'll refrain from reviewing the rest of the PR until that's commented on, as is it appears most of the change relies on passing state through in that fashion. If the team is ok with it, I am, too.)

I'm ok in general with the proposed design of passing the delegate (as an `object`) in the AsyncState property.

Please review the rest of the PR.

File: src/System.Net.Http/src/System/Net/Http/HttpClient.cs [hunk 2]
```diff
@@ -156,41 +156,124 @@ public Task<Stream> GetStreamAsync(string requestUri)
...
+                var innerReadAs = (Func<HttpContent, Task<string>>)innerTcs.Task.AsyncState;
+                if (HandleRequestFaultsAndCancelation(requestTask, innerTcs))
                 {
                     return;
                 }
                 HttpResponseMessage response = requestTask.Result;
                 if (response.Content == null)
                 {
-                    tcs.TrySetResult(defaultValue);
+                    innerTcs.TrySetResult(string.Empty);
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/HttpClient.cs]: I expect it's relatively uncommon for the content to be null?  Seems a bit unfortunate that we appear to be triplicating this whole method primarily due to needing a different default value here for each of the diferent types.  I haven't tried this, but alternative could be to just get a value based on the type being used and keeping a single implementation, e.g.
```C#
T result = 
    typeof(T) == typeof(string) ? (T)(object)string.Empty :
    typeof(T) == typeof(Stream) ? (T)(object)Stream.Null :
    typeof(T) == typeof(byte[]) ? (T)(object)Array.Empty<byte>() :
    default(T);
innerTcs.TrySetResult(result);
```
Alternatively, a static dictionary from type to default value could be maintained, prepopulated with all of these defaults.

**Comment by stephentoub** [comment]
Overall, the change looks fine to me and does remove a bunch of delegate and closure allocation.

I am hesitant, however, wondering if this is all temporary.  The current implementation is heavily based on ContinueWith, with a single operation represented as a chain of multiple continuations.  While this change definitely helps avoid a bunch of allocation costs associated with those continuations, I expect it would be much better to just rewrite a lot of this to use async/await instead; while a single ContinueWith can have some performance benefits when compared with an async method that has a single await in it, once you start adding more continuations to the same logical operation (as would be the case here), the performance benefits of async/await really start to kick in.  Switching to async/await, most (probably not all) of the changes made in this PR will be replaced.

So, I'm ok with the change, as I does appear to represent an improvement performance-wise (though I don't love the code duplication, as commented separately), and would be ok seeing it go in, but we may want to consider holding off on this and just doing what I expect will be the more impactful, performance-improving, and code-shrinking change of moving most of the implementation to using async/await.

Separately, with any change like this, it'd be helpful to understand what the reductions actually were, e.g. by using a memory profiler to understand how memory usage decreased, by using a throughput profiler to understand if there were any immediate throughput benefits, etc.

File: src/System.Net.Http/src/System/Net/Http/HttpClient.cs (hunk 2 above)
Comment (Maxwe11) [inline src/System.Net.Http/src/System/Net/Http/HttpClient.cs]: @stephentoub lambda in generic scope also could be cached?

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/HttpClient.cs]: With Roslyn, yes, it should be able to.  Pre-Roslyn, that can be handled by changing the method from being a generic method to being a non-generic method in a generic class.

**Comment by davidsh** [comment]
/cc @SidharthNabar 

As @stephentoub pointed out here and in the first PR review when System.Net.Http was put onto GitHub, many of the .ContinueWith() patterns can be improved to use async/await.   The Http code was written originally for .NET 4.5 (about 3+ years ago) and the Task/async patterns were still evolving.

I have concerns about the code duplication here that Stephen mentions.  Right now, there are only 3 Get*Async() helper methods for the HttpClient class.  But we envision expanding that to things like GetXmlAync(), GetJsonAsync(), etc.   There have been extension methods written as such already and shipped in NuGet packages by the ASP.NET team for example.  So, we see folding in those useful methods as part of moving the System.Net.Http, 4.0.0,0 contract forward for the next revision.   So, any changes we make here, we would want think about a clean and extendable design as well.

Additionally, we don't yet have all of our test code on GitHub.  We only have a fraction of the total test collateral.  Many of the Functional and multi-machine tests we have would be critical to fully validating more invasive changes for HttpClient.  We are working toward getting these tests onboarded over the next few months.

Improving performance for HttpClient is a high priority for us and we welcome all great feedback and ideas on this. I think it might be useful to get some performance testing of this change as a guideline for us.  But we might want to delay this specifc PR until we do the other more extensive async/await improvements.

**Comment by Maxwe11** [comment]
Changes to HttpClient which introduce code duplication excluded from PR.
HttpClient refactored with async/await #2220

**Comment by Maxwe11** [comment]
@dotnet-bot test this please

**Comment by jasonwilliams200OK** [comment]
@Maxwe11, you are fast! :+1:

**Comment by stephentoub** [comment]
Sorry for the delay.  The trimmed back revision of this LGTM.

**Comment by davidsh** [comment]
I will need to put this change thru our FunctionalTests (not yet on GitHub) before this can merged.

**Comment by davidsh** [comment]
LGTM

**Pull request merged**